library/public/camera.h

```diff
@@ -0,0 +1,93 @@
+/**
+ * @class   camera
+ * @brief   Abstract class to control a camera in a window
+ *
+ * A class to control a camera in a window.
+ * There is in essence three API.
+ * The Position/FocalPoint/ViewUp API, the ViewMatrix API and the camera movements API.
+ * They can be mixed by using ViewMatrix API with the other APIs but it can result in unexpected
+ * results.
+ * When using the Position/FocalPoint/ViewUp API or the camera movements API,
+ * the view up is automatically orthogonalized.
+ */
+
+#ifndef f3d_camera_h
+#define f3d_camera_h
+
+#include "export.h"
+
+#include <array>
+#include <string>
+
+namespace f3d
+{
+class F3D_EXPORT camera
+{
+public:
+  typedef std::array<double, 3> vector3_t;
+  typedef std::array<double, 16> matrix4_t;
+
+  //@{
+  /**
+   * Set/Get the camera parameters.
+   */
+  virtual camera& setPosition(const vector3_t& pos) = 0;
+  virtual vector3_t getPosition() = 0;
+  virtual void getPosition(vector3_t& pos) = 0;
+  virtual camera& setFocalPoint(const vector3_t& foc) = 0;
+  virtual vector3_t getFocalPoint() = 0;
+  virtual void getFocalPoint(vector3_t& foc) = 0;
+  virtual camera& setViewUp(const vector3_t& up) = 0;
+  virtual vector3_t getViewUp() = 0;
+  virtual void getViewUp(vector3_t& up) = 0;
+  virtual camera& setViewAngle(const double& angle) = 0;
+  virtual double getViewAngle() = 0;
+  virtual void getViewAngle(double& angle) = 0;
+  //@}
+
+  //@{
+  /**
+   * Set/Get the camera view matrix. Please note that when using the view matrix API, using the
+   * other camera manipulation API may results in unexepected results due to normalization.
+   */
+  virtual camera& setViewMatrix(const matrix4_t& matrix) = 0;
+  virtual matrix4_t getViewMatrix() = 0;
+  virtual void getViewMatrix(matrix4_t& matrix) = 0;
+  //@}
+
+  //@{
+  /**
+   * Standard camera manipulation methods.
+   */
+  virtual camera& dolly(double val) = 0;
+  virtual camera& roll(double angle) = 0;
```

Comment:
You should consider doing some kind of [strong typing](https://www.fluentcpp.com/2016/12/08/strong-types-for-strong-interfaces/) here and explicitly calling out Radians or Degrees as part of the function signature.

As it stands I do not know if this takes degrees or radians, and even if the documentation told me, there is still a lot more chance for errors taking a raw double than taking a class where you get type safety.

Reply:
that would be interesting to add, but not only, in many more locations, wdyt @Meakk ?

In the meantime, Ill just improve the documentation.

Reply:
I added a typedef but defining the templates as proposed in your link will not be in the alpha. Will we discuss if we want to do it,

Reply:
Partially adressed in https://github.com/f3d-app/f3d/pull/355, keeping this open for now.

Reply:
https://github.com/f3d-app/f3d/issues/361